tests/test_ci/TestControllers.py

```diff
@@ -487,3 +695,182 @@ def test_webhook_prerelease(self, mock_request):
+    @mock.patch('mod_ci.controllers.queue_test')
+    @mock.patch('mod_ci.controllers.GitHub')
+    @mock.patch('mod_ci.controllers.GeneralData')
+    def test_webhook_push_valid(self, mock_gd, mock_github, mock_queue_test, mock_request):
+        """
+        Test webhook triggered with push event without 'after' in payload.
+        """
+        data = {'after': 'abcdefgh'}
+        sig = generate_signature(str(json.dumps(data)).encode('utf-8'), g.github['ci_key'])
+        headers = generate_git_api_header('push', sig)
+        # one of ip address from github webhook
+        wsgi_environment = {'REMOTE_ADDR': '192.30.252.0'}
+        with self.app.test_client() as c:
+            response = c.post(
+                '/start-ci', environ_overrides=wsgi_environment,
+                data=json.dumps(data), headers=headers)
+
+        mock_gd.query.filter.assert_called()
+        mock_github.assert_called_once()
+        mock_queue_test.assert_called_once()
+
+    @mock.patch('mod_ci.controllers.Test')
+    @mock.patch('requests.get', side_effect=mock_api_request_github)
+    def test_webhook_pr_closed(self, mock_request, mock_test):
+        """
+        Test webhook triggered with pull_request event with closed action.
+        """
+        class MockTest:
+            def __init__(self):
+                self.id = 1
+
+        mock_test.query.filter.return_value.all.return_value = [MockTest()]
+
+        data = {'action': 'closed',
+                'pull_request': {'number': '1234'}}
+        sig = generate_signature(str(json.dumps(data)).encode('utf-8'), g.github['ci_key'])
+        headers = generate_git_api_header('pull_request', sig)
+        # one of ip address from github webhook
+        wsgi_environment = {'REMOTE_ADDR': '192.30.252.0'}
+        with self.app.test_client() as c:
+            response = c.post(
+                '/start-ci', environ_overrides=wsgi_environment,
+                data=json.dumps(data), headers=headers)
+
+        mock_test.query.filter.assert_called_once()
+
+    @mock.patch('mod_ci.controllers.BlockedUsers')
+    @mock.patch('mod_ci.controllers.GitHub')
+    @mock.patch('mod_ci.controllers.queue_test')
+    @mock.patch('requests.get', side_effect=mock_api_request_github)
+    def test_webhook_pr_opened_blocked(self, mock_request, mock_queue_test, mock_github, mock_blocked):
+        """
+        Test webhook triggered with pull_request event with opened action for blocked user.
+        """
+        class MockTest:
+            def __init__(self):
+                self.id = 1
+
+        data = {'action': 'opened',
+                'pull_request': {'number': '1234', 'head': {'sha': 'abcd1234'}, 'user': {'id': 'test'}}}
+        sig = generate_signature(str(json.dumps(data)).encode('utf-8'), g.github['ci_key'])
+        headers = generate_git_api_header('pull_request', sig)
+        # one of ip address from github webhook
+        wsgi_environment = {'REMOTE_ADDR': '192.30.252.0'}
```

Comment:
I see quite some repetetion in here (aside from the data)... Maybe extract the common parts to a single method to reduce the amount of duplicate code?

Reply:
fixed in d068c47a924f02d8c93316c1d87d5e993ff7e325